jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java

```diff
@@ -119,11 +119,23 @@ public boolean isSendError()
         return _state.isSendError();
     }
 
-    protected HttpInput newHttpInput(HttpChannelState state)
+    private HttpInput newHttpInput(HttpChannelState state)
     {
         return new HttpInput(state);
     }
 
+    /**
+     * @return true if content is immediately available.
+     */
+    public abstract boolean needContent();
```

Comment:
Needs an explanation of what exactly the contract is, because `produceContent()` returns `null` if there is no content, so why would you need `needContent()` - I'm guessing to implement `available()`, but needs to be made clear.

Reply:
EXACTLY!!!!!!!!!!!! `needContent` and `produceContent` are key parts of the new contract between HttpInput and HttpChannel so that we don't need to extend HttpInput on a a per transport/protocol basis.     

This is the contract that needs a lot more rigour in both documentation and implementation.

As I have said elsewhere, it is not sufficient for the response to a needContent to just be a wakeup call that can be called for a thousand other reasons.   It needs  to be a precise success of failure callback.   I really think it should be:
```suggestion
    public abstract boolean needContent(Callback callback);
```

If some error occurs when there is not a callback, then it needs to be delivered differently because there is no async or blocking operation for it to be delivered to.

Reply:
I agree that the contract currently is too loosely defined, but I think a good dose of javadoc should solve this.

Regarding the new `needContent(Callback)` signature, I find the idea interesting. I'm going to try it out to see if it would improve clarity.

Reply:
Hopefully the javadoc is now clear enough about what the contract is.

Reply:
maybe a note about that if a failure happens, then `onContentProduceable` will be called and an ErrorContent will return the error on the next call to `produceContent`

Reply:
Added javadoc explaining the failure behavior.